### PR TITLE
fix: incorrect unformat not format value

### PR DESCRIPTION
**Pull request by Scrum**

### CODE REVIEW

**Comment by Scrum** [comment]
see https://github.com/braintree/restricted-input/issues/58#issuecomment-494038672

File: lib/formatter/index.js [hunk 1]
```diff
@@ -64,16 +64,24 @@ Formatter.prototype.unformat = function (options) {
...
   var end = options.selection.end;
   var unformattedString = '';
+  var formatCompensation = 0;
 
   for (i = 0; i < this.pattern.length; i++) {
     patternChar = this.pattern[i];
 
-    if (!patternChar.isPermaChar && options.value[i] != null && patternChar.value.test(options.value[i])) {
-      unformattedString += options.value[i];
+    if (
```
Comment (crookedneighbor) [inline lib/formatter/index.js]: This logic is getting gnarly. Let's pull it out into it's own function.

File: test/unit/lib/formatter/index.js [hunk 2]
```diff
@@ -218,6 +218,18 @@ describe('Formatter', function () {
 
       expect(reformatted).to.deep.equal(options);
     });
+
+    it('can unformat not format value', function () {
```
Comment (crookedneighbor) [inline test/unit/lib/formatter/index.js]: can you rephrase this? I'm not sure I understand what it's testing.

File: test/unit/lib/formatter/index.js [hunk 3]
```diff
@@ -218,6 +218,18 @@ describe('Formatter', function () {
...
+      var formatter = new Formatter('{{9999}} {{9999}}');
+      var options = {
+        selection: {start: 6, end: 6},
+        value: '123456'
+      };
+
+      var unformatted = formatter.unformat(options);
```
Comment (crookedneighbor) [inline test/unit/lib/formatter/index.js]: For clarity, can we include an assertion about what the formatted value looks like?

File: test/unit/lib/formatter/index.js (hunk 2 above)
Comment (Scrum) [reply test/unit/lib/formatter/index.js]: change to `should return a unformatted value`. 
Let me know if this is not very suitable / understandable.

File: lib/formatter/index.js (hunk 1 above)
Comment (Scrum) [reply lib/formatter/index.js]: passed into function `isPatterntChar`

File: test/unit/lib/formatter/index.js (hunk 3 above)
Comment (Scrum) [reply test/unit/lib/formatter/index.js]: > For clarity, can we include an assertion about what the formatted value looks like?

Modified the test, let me know if I did not what you planned

**Comment by Scrum** [comment]
> We'll also need an integration test as well so we can fully test this behavior.

I will try to deal with this, but as I said earlier ruby is not the language that I know well

**Comment by Scrum** [comment]
@crookedneighbor tagging you just in case you don’t see this, otherwise.

**Comment by crookedneighbor** [comment]
I'll try to review this again by the end of the week.

**Comment by Scrum** [comment]
@crookedneighbor ping

File: test/support/index.html [hunk 4]
```diff
@@ -126,6 +131,11 @@
         pattern: input.getAttribute('data-pattern') || '{{9999}} {{9999}} {{9999}} {{9999}}'
       });
 
+      if (input.id === 'credit-card-unformatted-with-unformated-value') {
```
Comment (crookedneighbor) [inline test/support/index.html]: The id for the new input you created is `credit-card-number-with-unformated-value`

Comment (Scrum) [reply test/support/index.html]: fix

**Comment by Scrum** [comment]
> So, if I'm reading this correctly, this only matters when programmatically setting the value, and then changing the pattern? Or is there another case where this causes a problem?

That's right, I haven't found any other cases yet.

**Comment by Scrum** [comment]
@crookedneighbor Something went wrong with integration tests, unfortunately I can't restart the tests

**Comment by crookedneighbor** [comment]
The integration tests won't work for you, since they include encrypted environmental variables. Travis can only run them with the env variables when it's a PR from a branch off the main repo (otherwise there'd be a danger of leaking the variables).

**Comment by Scrum** [comment]
@crookedneighbor ping ?

**Comment by crookedneighbor** [comment]
I'll be out of office until Tuesday, so I won't be able to look at this again until next week.

**Comment by Scrum** [comment]
@crookedneighbor ping again

**Comment by Scrum** [comment]
@crookedneighbor ping

**Comment by crookedneighbor** [comment]
Taking a look at this again.

**Comment by crookedneighbor** [comment]
I've looked over this again and again, and I just don't think it's worth adding right now.

IMO, this is outside the scope of how restricted input functions. It cares about the keystrokes from the user. I'm not concerned with setting the value programmatically. (though, even in that case, since you're setting the value, you'll know ahead of time what the pattern should be can can simply set the pattern first instead of after the value is set)

**Comment by Scrum** [comment]
This fix fixes the wrong `unformat` because it does not take into account that the value may be unformatted this is not only an example in #59 it may be trite insert like copy/paste, yes, in this case, you know the format but do not take into account what's in `unformat` the unformatted value falls and because of this, the characters from the inserted value disappear

**Comment by crookedneighbor** [comment]
copy/paste is handled by a specific copy/paste handler. So I don't think the issue you are describing is reproducible with copy paste. 

The only way I could reproduce it is by setting a pattern that was shorter than the upcoming value, programmatically setting the value and then immediately setting the pattern. It's such an edge case, that I don't think it's worth complicating the source code to account for it (especially when it can be solved simply by setting the pattern ahead of time)
